src/java.base/share/classes/java/lang/runtime/Carriers.java

```diff
@@ -87,899 +82,106 @@
-     * the accessors will always be {@link Object} }
-     *
-     * @param methodType  {@link MethodType} whose parameter types supply the shape of the
-     *                    carrier's components
-     */
-    public static List<MethodHandle> components(MethodType methodType) {
-        return CarrierFactory
-                .of(methodType)
-                .components()
-                .stream()
-                .map(c -> c.asType(c.type().changeParameterType(0, Object.class)))
-                .toList();
+    private static Object constructor(Object... args) {
+        return new Carrier(args);
     }
 
     /**
      * {@return a component accessor {@link MethodHandle} for component {@code i} of the
      * carrier representing {@code methodType}. The receiver type of the accessor will always
      * be {@link Object} }
      *
-     * @param methodType  {@link MethodType} whose parameter types supply the shape of the
-     *                    carrier's components
-     * @param i           component index
+     * @param lookup Represents a lookup context with the accessibility
+     *               privileges of the caller.  When used with {@code invokedynamic},
+     *               this is stacked automatically by the VM.
+     * @param invocationName unused
+     * @param invocationType The invocation type of the {@code CallSite} with one parameter,
+     *                       a reference type, and type that matches the i-th parameter type
+     *                       from {@code methodType} as a return type.
+     * @param carrierShape {@link MethodType} whose parameter types supply the shape of the
+     *                     carrier's components
+     * @param i            component index
      *
      * @throws IllegalArgumentException if {@code i} is out of bounds
      */
-    public static MethodHandle component(MethodType methodType, int i) {
-        MethodHandle component = CarrierFactory.of(methodType).component(i);
-        component = component.asType(component.type().changeParameterType(0, Object.class));
-        return component;
+    public static CallSite component(MethodHandles.Lookup lookup,
+                                         String invocationName,
+                                         MethodType invocationType,
+                                         MethodType carrierShape,
+                                         int i) {
+        Class<?> componentType = carrierShape.parameterType(i);
+        if (invocationType.parameterCount() != 1
+            || (!invocationType.returnType().equals(componentType))
+            || !invocationType.parameterType(0).equals(Object.class))
+            throw new IllegalArgumentException("Illegal invocation type " + invocationType);
+        return new ConstantCallSite(MethodHandles.insertArguments(COMPONENT_GETTER, 0, i).asType(MethodType.methodType(componentType, Object.class)));
```

Comment:
Since you're inserting the first argument I'd go for `COMPONENT_GETTER.bindTo(i)`

Reply:
I don't think that's possible - the parameter is an `int`, and `bindTo` requires the first parameter to be a reference.

Reply:
Ah, that's a good observation :+1: